rir/src/compiler/pir/type.cpp

```diff
@@ -123,5 +131,22 @@ void PirType::merge(const ObservedValues& other) {
         merge(record.sexptype);
     }
 }
+
+bool PirType::hasInstance(SEXP val) const {
+    if (isRType()) {
+        if (TYPEOF(val) == PROMSXP)
+            return maybePromiseWrapped() || maybeLazy() ||
+                   PirType(RType::prom).isA(*this);
+        if (LazyEnvironment::cast(val))
+            return PirType(RType::env).isA(*this);
+        return PirType(val).isA(*this | RType::missing);
```

Comment:
why `| missing`? doesn't make sense. if val is the missing value, then the type should allow for missingness, otherwise it is an error.

Reply:
can you please also check scalar and objectness?

Reply:
Wouldn't these already be handled though? e.g. `PirType(<object>).isA(<non-object type>) => <object type>.isA(<non-object type>) => false`